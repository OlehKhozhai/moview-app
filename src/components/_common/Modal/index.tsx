import React from 'react';
import { createPortal } from 'react-dom';

import styles from './styles.module.scss';

type ModalProps = {
  isOpen: boolean;
  children: React.ReactNode;
  onClose: () => void;
};

const Modal: React.FC<ModalProps> = ({ children, isOpen, onClose }) => {
  const modalNode = document.getElementById('modal');
  const component = (
    <div className={styles.root}>
      <div className={styles.overlay}>
        <div className={styles.modal}>
          <button className={styles.closeBtn} onClick={onClose}>
            +
          </button>
          {children}
        </div>
      </div>
    </div>
  );

  React.useEffect(() => {
    if (isOpen) {
      document.body.style.overflowY = 'hidden';
    }

    return () => {
      document.body.style.overflowY = 'auto';
    };
  }, [isOpen]);

  return modalNode && isOpen ? createPortal(component, modalNode) : null;
};

export default React.memo(Modal);
