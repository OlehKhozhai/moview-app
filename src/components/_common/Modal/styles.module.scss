@import 'assets/styles/mixins/colors';

.root {
  display: flex;
  align-items: center;
  justify-content: center;
  background-color: $primary-light-bg;
  position: fixed;
  top: 0;
  bottom: 0;
  right: 0;
  left: 0;
  z-index: 100;
}

.overlay {
  width: 100%;
  max-height: 100vh;
  overflow-y: auto;
}

.modal {
  padding: 80px;
  background-color: #232323;
  color: $primary-color;
  position: relative;
  border-radius: 4px;
  overflow-y: auto;
  max-width: 40vw;
  margin: 50px auto;
}

.closeBtn {
  position: absolute;
  right: 35px;
  top: 0;
  cursor: pointer;
  border: none;
  color: $primary-color;
  font-weight: 400;
  font-size: 60px;
  background-color: transparent;
  transform: rotate(135deg);
}
