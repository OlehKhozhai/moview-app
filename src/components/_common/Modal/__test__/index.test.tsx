import React from 'react';
import ReactDOM from 'react-dom';
import { create } from 'react-test-renderer';

import Modal from '../index';

describe('Modal component', () => {
  const modalRoot = window.document.createElement('div');
  const body = window.document.querySelector('body');
  modalRoot.setAttribute('id', 'modal');

  beforeAll(() => {
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    ReactDOM.createPortal = jest.fn((element: any) => element);
  });

  beforeEach(() => {
    body?.appendChild(modalRoot);
  });

  afterEach(() => {
    body?.removeChild(modalRoot);
  });

  test('should render Modal component by default', () => {
    const component = create(
      <Modal isOpen={true} onClose={() => undefined}>
        Modal content
      </Modal>
    );

    expect(component.toJSON()).toMatchSnapshot();
  });

  test('should render Modal component with value of isOpen is false  ', () => {
    const component = create(
      <Modal isOpen={false} onClose={() => undefined}>
        Modal content
      </Modal>
    );

    expect(component.toJSON()).toMatchSnapshot();
  });

  test('should trigger onClose method', () => {
    const handleOnClose = jest.fn();
    const component = create(
      <Modal isOpen={true} onClose={handleOnClose}>
        Modal content
      </Modal>
    );

    component.root.props.onClose();

    expect(handleOnClose).toHaveBeenCalled();
  });
});
