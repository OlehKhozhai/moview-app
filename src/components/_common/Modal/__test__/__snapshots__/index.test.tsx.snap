// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Modal component should render Modal component by default 1`] = `
<div
  className="root"
>
  <div
    className="overlay"
  >
    <div
      className="modal"
    >
      <button
        className="closeBtn"
        onClick={[Function]}
      >
        +
      </button>
      Modal content
    </div>
  </div>
</div>
`;

exports[`Modal component should render Modal component with value of isOpen is false   1`] = `null`;
